x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,275 @@
+	return fmt.Sprintf(`
+          {"channel_id": "channel-0", "denom": "%s", "quotas": [{"name":"%s", "duration": %d, "send_recv":[%d, %d]}] }
+    `, sdk.DefaultBondDenom, name, duration, send_precentage, recv_percentage)
+}
+
+// Tests
+
+// Test that Sending IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestSendTransferNoContract() {
+	one := sdk.NewInt(1)
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, one))
+}
+
+// Test that Receiving IBC messages works when the middleware isn't configured
+func (suite *MiddlewareTestSuite) TestReceiveTransferNoContract() {
+	one := sdk.NewInt(1)
+	suite.AssertReceiveSuccess(true, suite.NewValidMessage(false, one))
+}
+
+// Test rate limiting on sends
+func (suite *MiddlewareTestSuite) TestSendTransferWithRateLimiting() map[string]string {
+	// Setup contract
+	suite.chainA.StoreContractCode(&suite.Suite)
+	quotas := suite.BuildChannelQuota("weekly", 604800, 5, 5)
+	addr := suite.chainA.InstantiateContract(&suite.Suite, quotas)
+	suite.chainA.RegisterRateLimitingContract(addr)
+
+	// Setup sender chain's quota
+	osmosisApp := suite.chainA.GetOsmosisApp()
+
+	// Each user has 10% of the supply
+	supply := osmosisApp.BankKeeper.GetSupply(suite.chainA.GetContext(), sdk.DefaultBondDenom)
+	quota := supply.Amount.QuoRaw(20)
+	half := quota.QuoRaw(2)
+
+	// send 2.5% (quota is 5%)
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, half))
+
+	// send 2.5% (quota is 5%)
+	r, _ := suite.AssertSendSuccess(true, suite.NewValidMessage(true, half))
+
+	// Calculate remaining allowance in the quota
+	attrs := suite.ExtractAttributes(suite.FindEvent(r.GetEvents(), "wasm"))
+	used, _ := sdk.NewIntFromString(attrs["weekly_used_out"])
+	suite.Require().Equal(used, half.MulRaw(2))
+
+	// Sending above the quota should fail.
+	suite.AssertSendSuccess(false, suite.NewValidMessage(true, sdk.NewInt(1)))
+	return attrs
+}
+
+// Test rate limits are reset when the specified time period has passed
+func (suite *MiddlewareTestSuite) TestSendTransferReset() {
+	// Same test as above, but the quotas get reset after time passes
+	attrs := suite.TestSendTransferWithRateLimiting()
+	parts := strings.Split(attrs["weekly_period_end"], ".") // Splitting timestamp into secs and nanos
+	secs, _ := strconv.ParseInt(parts[0], 10, 64)
+	nanos, _ := strconv.ParseInt(parts[1], 10, 64)
+	resetTime := time.Unix(secs, nanos)
+
+	// Move both chains one block
+	suite.chainA.NextBlock()
+	suite.chainA.SenderAccount.SetSequence(suite.chainA.SenderAccount.GetSequence() + 1)
+	suite.chainB.NextBlock()
+	suite.chainB.SenderAccount.SetSequence(suite.chainB.SenderAccount.GetSequence() + 1)
+
+	// Reset time + one second
+	oneSecAfterReset := resetTime.Add(time.Second)
+	suite.coordinator.IncrementTimeBy(oneSecAfterReset.Sub(suite.coordinator.CurrentTime))
+
+	// Sending should succeed again
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, sdk.NewInt(1)))
```

Comment:
Should we error-check here?

Reply:
`AssertSendSuccess` does error checking internally